Android/advanced-kotlin/app/src/main/java/com/example/immersivereadersdk/MainActivity.kt

```diff
@@ -0,0 +1,215 @@
+        val immersiveReaderButton = findViewById<Button>(R.id.LaunchImmersiveReaderButton)
+        immersiveReaderButton.setOnClickListener { GlobalScope.launch { handleLoadImmersiveReaderWebView() } }
+    }
+
+    private suspend fun handleLoadImmersiveReaderWebView() {
+        val exampleActivity = this
+        val subdomain = dotEnv["SUBDOMAIN"]
+        val irTitle = findViewById<TextView>(R.id.Title)
+        val irText = findViewById<TextView>(R.id.Content)
+        val chunk = Chunk(irText.text.toString(), "en", "text/plain")
+        val chunks = listOf(chunk)
+        val data = Content(irTitle.text.toString(), chunks)
+        val options = Options("ImmersiveReader-Exit","en", 0)
+        var token: String
+
+        runBlocking{
+            val resp = async { getImmersiveReaderTokenAsync() }
+            token = resp.await()
+            val jsonResp = JSONObject(token)
+            loadImmersiveReaderWebView(exampleActivity, jsonResp.getString("access_token"), subdomain, data, options)
+        }
+    }
+
+    private suspend fun getImmersiveReaderTokenAsync(): String {
+        return getToken()
+    }
+
+    @Throws(IOException::class)
+    fun getToken(): String {
+        val clientId = dotEnv["CLIENT_ID"]
+        val clientSecret = dotEnv["CLIENT_SECRET"]
+        val tenantId = dotEnv["TENANT_ID"]
+        val tokenUrl = URL("https://login.windows.net/$tenantId/oauth2/token")
+        val form = "grant_type=client_credentials&resource=https://cognitiveservices.azure.com/&client_id=$clientId&client_secret=$clientSecret"
+
+        val connection = tokenUrl.openConnection() as HttpURLConnection
+        connection.requestMethod = "POST"
+        connection.setRequestProperty("content-type", "application/x-www-form-urlencoded")
+        connection.doOutput = true
+
+        val writer = DataOutputStream(connection.outputStream)
+        writer.writeBytes(form)
+        writer.flush()
+        writer.close()
+
+        val responseCode = connection.responseCode
+
+        if (responseCode == HTTP_OK) {
+            val readerIn = BufferedReader(InputStreamReader(connection.inputStream))
+            var inputLine = readerIn.readLine()
+            val response = StringBuffer()
+
+            do {
+                response.append(inputLine)
+            } while (inputLine.length < 0)
+            readerIn.close()
+
+            // Return token
+            return response.toString()
+        } else {
+            val responseError = Error(code = "BadRequest", message = "There was an error getting the token.")
+            throw IOException(responseError.toString())
+        }
+    }
+
+    private lateinit var contextualWebView: WebView
+
+    data class Content(var title: String,
+                       var chunks: List<Chunk>)
+
+    data class Chunk(var content: String,
+                     var lang: String,
+                     var mimeType: String)
+
+    data class Options(var exitCallback: String,

```

Comment:
Sure can, will add them - do we have a list of supported options somewhere?

Reply:
https://docs.microsoft.com/en-us/azure/cognitive-services/immersive-reader/reference#options 

Why do we need these classes? if we keep these classes then we'll have to update them each time we add an option. Is it possible to just deserialize a dictionary for readability and maintainability?

Reply:
@JavaScriptInterface will impact the readability for sure (It will be complex as the dictionaries will need to be defined in JavaScript and used by Kotlin via a function). Though the maintainability will be better. Will investigate.

Reply:
@dmcallis and I had a quick discussion about this and these data classes will be part of the Android SDK package, as will the loadImmersiveReaderWebView function. At the moment they're in the sample app - but should be part of an imported package.

Reply:
The good thing about that is we do the right thing by our users and we maintain these classes - not them. The sample app will then deserialize a dictionary of the classes we provide for readability and maintainability.

Reply:
> https://docs.microsoft.com/en-us/azure/cognitive-services/immersive-reader/reference#options
> Why do we need these classes? if we keep these classes then we'll have to update them each time we add an option. Is it possible to just deserialize a dictionary for readability and maintainability?

If you serialize a dictionary, you still need to keep that dictionary updated as we add more properties to the SDK. I don't think it matters which format we use, classes or dynamic object or dictionary, with respect to maintainability for us. 

But even if it is more work for us to maintain, I think giving developers the Content class and saying: "create an object of this type and pass it to us" makes for a better SDK than asking them to create their own dynamic object or dictionary.

Reply:
Will leave as is for now. Experimented using HashMaps and the perf was not good.